polkadot/consensus/src/shared_table/mod.rs

```diff
@@ -0,0 +1,566 @@
+}
+
+// A shared table object.
+struct SharedTableInner {
+	table: Table<TableContext>,
+	proposed_digest: Option<Hash>,
+	checked_validity: HashSet<Hash>,
+	checked_availability: HashSet<Hash>,
+	trackers: Vec<IncludabilitySender>,
+}
+
+impl SharedTableInner {
+	// Import a single statement. Provide a handle to a table router and a function
+	// used to determine if a referenced candidate is valid.
+	fn import_statement<R: TableRouter, C: FnMut(Collation) -> bool>(
+		&mut self,
+		context: &TableContext,
+		router: &R,
+		statement: table::SignedStatement,
+		statement_source: StatementSource,
+		check_candidate: C,
+	) -> StatementProducer<
+		<R::FetchCandidate as IntoFuture>::Future,
+		<R::FetchExtrinsic as IntoFuture>::Future,
+		C,
+	> {
+		// this blank producer does nothing until we attach some futures
+		// and set a candidate digest.
+		let received_from = match statement_source {
+			StatementSource::Local => return Default::default(),
+			StatementSource::Remote(from) => from,
+		};
+
+		let summary = match self.table.import_statement(context, statement, received_from) {
+			Some(summary) => summary,
+			None => return Default::default(),
+		};
+
+		self.update_trackers(&summary.candidate, context);
+
+		let local_id = context.local_id();
+
+		let is_validity_member = context.is_member_of(&local_id, &summary.group_id);
+		let is_availability_member =
+			context.is_availability_guarantor_of(&local_id, &summary.group_id);
+
+		let digest = &summary.candidate;
+
+		// TODO: consider a strategy based on the number of candidate votes as well.
+		// only check validity if this wasn't locally proposed.
+		let checking_validity = is_validity_member
+			&& self.proposed_digest.as_ref().map_or(true, |d| d != digest)
+			&& self.checked_validity.insert(digest.clone());
+
+		let checking_availability = is_availability_member
+			&& self.checked_availability.insert(digest.clone());
+
+		let work = if checking_validity || checking_availability {
+			match self.table.get_candidate(&digest) {
+				None => None, // TODO: handle table inconsistency somehow?
+				Some(candidate) => {
+					let fetch_block_data =
+						router.fetch_block_data(candidate).into_future().fuse();
+
+					let fetch_extrinsic = if checking_availability {
+						Some(
+							router.fetch_extrinsic_data(candidate).into_future().fuse()
+						)
+					} else {
+						None
+					};
+
+					Some(Work {
+						candidate_receipt: candidate.clone(),
+						fetch_block_data,
+						fetch_extrinsic,
+						evaluate: checking_validity,
+						check_candidate,
+					})
+				}
+			}
+		} else {
+			None
+		};
+
+		StatementProducer {
+			produced_statements: Default::default(),
+			work,
+		}
+	}
+
+	fn update_trackers(&mut self, candidate: &Hash, context: &TableContext) {
+		let includable = self.table.candidate_includable(candidate, context);
+		for i in (0..self.trackers.len()).rev() {
+			if self.trackers[i].update_candidate(candidate.clone(), includable) {
```

Comment:
This may obviously mess up the ordering in `self.trackers` but I suppose we don't care about the order anyway.

Reply:
yeah, it's fine to reorder in this case